Modules/blake2module.c

```diff
@@ -0,0 +1,925 @@
+            PyErr_Format(PyExc_ValueError,
+                "maximum person length is %d bytes",
+                (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_PERSONAL_BYTES : HACL_HASH_BLAKE2S_PERSONAL_BYTES));
+            goto error;
+        }
+        memcpy(personal_, person->buf, person->len);
+    }
+
+    /* Validate tree parameters. */
+    if (fanout < 0 || fanout > 255) {
+        PyErr_SetString(PyExc_ValueError,
+                "fanout must be between 0 and 255");
+        goto error;
+    }
+
+    if (depth <= 0 || depth > 255) {
+        PyErr_SetString(PyExc_ValueError,
+                "depth must be between 1 and 255");
+        goto error;
+    }
+
+    if (leaf_size > 0xFFFFFFFFU) {
+        PyErr_SetString(PyExc_OverflowError, "leaf_size is too large");
+        goto error;
+    }
+
+    if (is_blake2s(self->impl) && node_offset > 0xFFFFFFFFFFFFULL) {
+        /* maximum 2**48 - 1 */
+         PyErr_SetString(PyExc_OverflowError, "node_offset is too large");
+         goto error;
+     }
+
+    if (node_depth < 0 || node_depth > 255) {
+        PyErr_SetString(PyExc_ValueError,
+                "node_depth must be between 0 and 255");
+        goto error;
+    }
+
+    if (inner_size < 0 ||
+        (unsigned) inner_size > (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_OUT_BYTES : HACL_HASH_BLAKE2S_OUT_BYTES)) {
+        PyErr_Format(PyExc_ValueError,
+                "inner_size must be between 0 and is %d",
+                (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_OUT_BYTES : HACL_HASH_BLAKE2S_OUT_BYTES));
+        goto error;
+    }
+
+    /* Set key length. */
+    if ((key->obj != NULL) && key->len) {
+        if (key->len > (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_KEY_BYTES : HACL_HASH_BLAKE2S_KEY_BYTES)) {
+            PyErr_Format(PyExc_ValueError,
+                "maximum key length is %d bytes",
+                (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_KEY_BYTES : HACL_HASH_BLAKE2S_KEY_BYTES));
+            goto error;
+        }
+    }
+
+    // Unlike the state types, the parameters share a single (client-friendly)
+    // structure.
+
+    Hacl_Hash_Blake2b_blake2_params params = {
+        .digest_length = digest_size,
+        .key_length = key->len,
+        .fanout = fanout,
+        .depth = depth,
+        .leaf_length = leaf_size,
+        .node_offset = node_offset,
+        .node_depth = node_depth,
+        .inner_length = inner_size,
+        .salt = salt_,
+        .personal = personal_
+    };
+
+    switch (self->impl) {
+#if HACL_CAN_COMPILE_SIMD256
+        case Blake2b_256:
+            self->blake2b_256_state = Hacl_Hash_Blake2b_Simd256_malloc_with_params_and_key(&params, last_node, key->buf);
+            break;
+#endif
+#if HACL_CAN_COMPILE_SIMD128
+        case Blake2s_128:
+            self->blake2s_128_state = Hacl_Hash_Blake2s_Simd128_malloc_with_params_and_key(&params, last_node, key->buf);
+            break;
+#endif
```

Comment:
Use `{ }` even in cases:

```C
case ...: {
 ....
}
```

Reply:
That feels like overkill, we don't do it most anywhere else within CPython. that style is only used 1-2 dozen times while `case` appears without braces over a thousand.  If I read code doing this I'd expect there to be a reason such as a case-local variable declaration.